driver/generic/xpt2046.py

```diff
@@ -0,0 +1,123 @@
+        *ranges*: `(x_min,x_max),(y_min,y_max)` for raw coordinate readings; calibrated values might be provided.
+        *width*: width of the underyling screen in pixels, in natural (rot=0) orientation (0..*width* is the range for reported horizontal coordinate)
+        *height*: height of the underyling screen in pixels
+        *rot*: screen rotation (0: portrait, 1: landscape, 2: inverted portrait, 3: inverted landscape); the constants XPT2046_PORTRAIT, XPT2046_LANDSCAPE, XPT2046_INV_PORTRAIT, XPT2046_INV_LANDSCAPE may be used.
+        '''
+        self.buf = bytearray(3)
+        self.spi = spi
+        self.cs  = (machine.Pin(cs,machine.Pin.OUT) if isinstance(cs,int) else cs)
+        self.cs.value(1)
+        if bits not in (8,12): raise ValueError('Xpt2046.bits: must be 8 or 12 (not %s)'%str(bits))
+        self.conv=(Xpt2046.CONV_8_BIT if bits==8 else Xpt2046.CONV_12_BIT)
+        self.xy_range,self.dim,self.rot=ranges,(width,height),(rot%4)
+        self.xy_scale=[self.dim[ax]*1./(self.xy_range[ax][1]-self.xy_range[ax][0]) for ax in (0,1)]
+        self.xy_origin=[self.xy_range[ax][0] for ax in (0,1)]
+    def _raw2px(self,rxy):
+        'Convert raw coordinates to pixel coordinates'
+        x,y=[int(self.xy_scale[ax]*(rxy[ax]-self.xy_origin[ax])) for ax in (0,1)]
+        if   self.rot==0: return self.dim[0]-x,y
+        elif self.rot==1: return self.dim[1]-y,x
+        elif self.rot==2: return x,self.dim[1]-y
+        else:             return y,self.dim[0]-x
+    def _raw_pos(self):
+        'Read raw position; return value if within valid ranges (`__init__(ranges=...)`) or `None` if outside.'
+        ret=[0,0]
+        for ax,chan in [(0,Xpt2046.CHAN_X),(1,Xpt2046.CHAN_Y)]:
+            r=self._chanRead(chan)
+            if not self.xy_range[ax][0]<=r<=self.xy_range[ax][1]: return None
+            ret[ax]=r
+        return ret
+
+    def pos(self,N=10,attempts=20):
+        ''''
+        Get N position readings (limited by 20 attempts) and return mean position of valid readings.
+        If attempts are exhausted, return None.
+        '''
+        N,attempts=10,20
+        xx,yy,done=0,0,0
+        for _ in range(attempts):
+            if (r:=self._raw_pos()) is None: continue
+            xx+=r[0]; yy+=r[1]; done+=1
+            if done==N: break
+        else: return None
+        mx,my=xx*1./N,yy*1./N
+        return self._raw2px((mx,my))
+
+
+class Xpt2046(Xpt2046_hw):
+    lcd=None
+    spi=None
+    def indev_drv_read_cb(self, indev_drv, data):
+        # wait for DMA transfer (if any) before switchint SPI to 1 MHz
+        if self.lcd: self.lcd._rp2_wait_dma()
+        # print('.',end='')
+        if self.spiRate: self.spi.init(baudrate=1_000_000)
+        pos=self.pos()
+        if pos is None: data.state=0
+        else: (data.point.x,data.point.y),data.state=pos,1
+        # print('#',end='')
+        # switch SPI back to 24 MHz for the display
```

Comment:
I don't think you should assume the display is necessarily 24 MHz.

Reply:
This was an old comment which I removed; the new rate will be Xpt2046.spiRate.